### ISSUE TITLE
[NOT ISSUE] Do you have any plan to support Tensorflow Training on Graphene 

### ISSUE THREAD

**Issue by duanbing** [oscarlab/graphene#2127]
Since Tensorflow is so much widely used, I am looking for some solution to migrate the end-to-end training process into Enclave.  What I know is that tensorflow lite as a predictor is supported by LibOS like Graphene, and PPML on PyTorch is also workable here.   

Would be very appreciative of your replying!

**Comment by dimakuv** [comment]
TensorFlow is also supported by Graphene, we have reports that people successfully enabled their TensorFlow workloads in Graphene.

Unfortunately, we do not have an example in our repository yet. But you are encouraged to create your own example, test it and submit a PR!

**Comment by duanbing** [comment]
Sure. I will. Here I have some question to ask. I tested the without sgx PPML, now I try to run it with SGX=1. Unluckily I have an older kernel. 
```
4.14.81.bm.26-amd64
```
and, after I built and inserted gsgx module,  and got:
```
LD_SHOW_AUXV=1 /bin/true | grep AT_HWCAP2
AT_HWCAP2:            0x0
```
and 2 is not set, which be considered as a failure to install gsgx module?  and then when I compile graphene by 
```
make SGX=1
```
got errors:

```
In file included from ../../../include/arch/x86_64/Linux/pal_host-arch.h:18,
                 from ../../../include/arch/x86_64/pal-arch.h:46,
                 from ../../../include/pal/pal.h:90,
                 from pal_linux.h:14,
                 from sgx_framework.c:5:
sgx_framework.c: In function 'open_sgx_driver':
sgx_framework.c:20:49: error: 'GSGX_FILE' undeclared (first use in this function); did you mean 'STT_FILE'?
   20 |         g_gsgx_device = INLINE_SYSCALL(open, 3, GSGX_FILE, O_RDWR | O_CLOEXEC, 0);
...

I found GSGX_FILE is defined in `graphene-sgx-driver`,  as well as `./Pal/src/host/Linux-SGX/gsgx.h.in`.  right now I am still not familiar with source code of graphene, Could you  give me some tips to get through of it ?

**Comment by dimakuv** [comment]
> and 2 is not set, which be considered as a failure to install gsgx module? 

This doesn't mean much because you're on an older Linux kernel. You can check `dmesg` output; it must have a big warning message that GSGX kernel module is running in the kernel. This should be enough to make sure you can use FSGSBASE hardware capability.

>  found GSGX_FILE is defined in graphene-sgx-driver, as well as ./Pal/src/host/Linux-SGX/gsgx.h.in. right now I am still not familiar with source code of graphene, Could you give me some tips to get through of it?

Are you sure you built Graphene with the SGX backend? You need to do something like this: https://listed.to/@jarkko/21962/using-graphene-to-run-applications-ins.

There is a known problem when you rebuild the SGX backend from an incorrect state. So please do this before proceeding (to clean everything old in Graphene):
```
cd <graphene root dir>
make distclean
make SGX=1 distclean

# now you can re-build Graphene again, using the instructions in the above link
```

**Comment by duanbing** [comment]
Great!  I run the `bash` in SGX successfully according to your instruction above.  Now another issue occurred as below when I run the example of PPML: 
```
root@-:/app/graphene/Examples/pytorch# SGX=1 ./pal_loader ./pytorch pytorchexample.py
root@-:/app/graphene/Examples/pytorch# echo $?
2
```
And no `results.txt`  generated.

**Comment by duanbing** [comment]
When I run all the above example on Ubuntu18.04, all are workable.

**Comment by dimakuv** [comment]
What is the Docker image you're using? Is it based off Ubuntu 18.04? Looks like there are some standard paths that differ from Ubuntu 18.04 (all our examples are tested only on Ubuntu 18.04).

**Comment by duanbing** [comment]
@dimakuv  Got it， previous errors were reported from Ubuntu20.  It's ok now on Ubuntu18.

**Comment by mkow** [comment]
So, is there anything left or can we close this issue? Ubuntu 20.04 is not officially supported yet.

**Comment by duanbing** [comment]
Not yet I am afraid. I am still struggling to support TF training.  the manifest is [here](https://github.com/duanbing/graphene/blob/master/Examples/tensorflow-training/label_image.manifest.template). 

When I run `make SGX=1 DEBUG=1`, **hung**, and the last logs as below:
```
[P98777:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages", 0xb629c9c0) = 0x0
[P98777:T1:python3] ---- shim_stat("/usr/lib/python3/dist-packages", 0xb629c9c0) = 0x0
[P98777:T1:python3] ---- shim_ioctl(0, TCGETS, 0xb629f380) ...
[P98777:T1:python3] ---- return from shim_ioctl(...) = -38
```

**Comment by dimakuv** [comment]
@duanbing Unfortunately, it's hard to figure out anything from these 4 lines of the log. Could you copy-paste the last 100 lines? Do any other workloads work for you (just as a sanity check)?

**Comment by duanbing** [comment]
@dimakuv 
Examples like GCC, bash and pytorch all are work for me.  

```
...
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib", 0x8d2687d0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib/abc.py", 0x8d2683f0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib/abc.py", 0x8d268c90) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/importlib/__pycache__/abc.cpython-36.pyc", O_RDONLY|0x80000, 0000) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d268df0) = 0x0
[P99332:T1:python3] ---- shim_lseek(4, 0x0, SEEK_CUR) = 0x0
[P99332:T1:python3] ---- shim_fstat(4, 0x8d268e50) = 0x0
[P99332:T1:python3] ---- shim_read(4, 0x8bd214f0, 0x2c1e) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x2c1d
[P99332:T1:python3] ---- shim_read(4, 0x8bd2410d, 0x1) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x0
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/importlib/abc.py", O_RDONLY|0x80000, 0000) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d268df0) = 0x0
[P99332:T1:python3] ---- shim_lseek(4, 0x0, SEEK_CUR) = 0x0
[P99332:T1:python3] ---- shim_fstat(4, 0x8d268e50) = 0x0
[P99332:T1:python3] ---- shim_read(4, 0x8bd1b970, 0x2a1f) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x2a1e
[P99332:T1:python3] ---- shim_read(4, 0x8bd1e38e, 0x1) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x0
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib/abc.py", 0x8d268aa0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib/__pycache__", 0x8d268590) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/importlib/__pycache__/abc.cpython-36.pyc.2344733696", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0600) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d2687b0) = 0x0
[P99332:T1:python3] ---- shim_write(4, 0x8bdffbe0, 0x2c1d) ...
[P99332:T1:python3] ---- return from shim_write(...) = 0x2c1d
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_rename("/usr/lib/python3.6/importlib/__pycache__/abc.cpython-36.pyc.2344733696", "/usr/lib/python3.6/importlib/__pycache__/abc.cpython-36.pyc") = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib", 0x8d267060) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib/machinery.py", 0x8d266c80) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib/machinery.py", 0x8d267520) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/importlib/__pycache__/machinery.cpython-36.pyc", O_RDONLY|0x80000, 0000) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d267680) = 0x0
[P99332:T1:python3] ---- shim_lseek(4, 0x0, SEEK_CUR) = 0x0
[P99332:T1:python3] ---- shim_fstat(4, 0x8d2676e0) = 0x0
[P99332:T1:python3] ---- shim_read(4, 0x8bd97510, 0x3b7) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x3b6
[P99332:T1:python3] ---- shim_read(4, 0x8bd978c6, 0x1) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x0
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/importlib/machinery.py", O_RDONLY|0x80000, 0000) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d267680) = 0x0
[P99332:T1:python3] ---- shim_lseek(4, 0x0, SEEK_CUR) = 0x0
[P99332:T1:python3] ---- shim_fstat(4, 0x8d2676e0) = 0x0
[P99332:T1:python3] ---- shim_read(4, 0x8bd94b50, 0x34d) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x34c
[P99332:T1:python3] ---- shim_read(4, 0x8bd94e9c, 0x1) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x0
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib/machinery.py", 0x8d267330) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/importlib/__pycache__", 0x8d266e20) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/importlib/__pycache__/machinery.cpython-36.pyc.2343645592", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0600) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d267040) = 0x0
[P99332:T1:python3] ---- shim_write(4, 0x8bd25bc0, 0x3b6) ...
[P99332:T1:python3] ---- return from shim_write(...) = 0x3b6
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_rename("/usr/lib/python3.6/importlib/__pycache__/machinery.cpython-36.pyc.2343645592", "/usr/lib/python3.6/importlib/__pycache__/machinery.cpython-36.pyc") = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6", 0x8d269000) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/contextlib.py", 0x8d268c20) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/contextlib.py", 0x8d2694c0) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/__pycache__/contextlib.cpython-36.pyc", O_RDONLY|0x80000, 0000) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d269620) = 0x0
[P99332:T1:python3] ---- shim_lseek(4, 0x0, SEEK_CUR) = 0x0
[P99332:T1:python3] ---- shim_fstat(4, 0x8d269680) = 0x0
[P99332:T1:python3] ---- shim_read(4, 0x8bd22460, 0x2b97) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x2b96
[P99332:T1:python3] ---- shim_read(4, 0x8bd24ff6, 0x1) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x0
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/contextlib.py", O_RDONLY|0x80000, 0000) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d269620) = 0x0
[P99332:T1:python3] ---- shim_lseek(4, 0x0, SEEK_CUR) = 0x0
[P99332:T1:python3] ---- shim_fstat(4, 0x8d269680) = 0x0
[P99332:T1:python3] ---- shim_read(4, 0x8bd1b970, 0x336b) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x336a
[P99332:T1:python3] ---- shim_read(4, 0x8bd1ecda, 0x1) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x0
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/contextlib.py", 0x8d2692d0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/__pycache__", 0x8d268dc0) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/__pycache__/contextlib.cpython-36.pyc.2344734480", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0600) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d268fe0) = 0x0
[P99332:T1:python3] ---- shim_write(4, 0x8bdffb60, 0x2b96) ...
[P99332:T1:python3] ---- return from shim_write(...) = 0x2b96
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_rename("/usr/lib/python3.6/__pycache__/contextlib.cpython-36.pyc.2344734480", "/usr/lib/python3.6/__pycache__/contextlib.cpython-36.pyc") = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages", 0x8d26a940) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages", 0x8d26a0e0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages", 0x8d26a8a0) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/local/lib/python3.6/dist-packages", O_RDONLY|0x90800, 0000) = 0x4
[P99332:T1:python3] ---- shim_fstat(4, 0x8d26a7c0) = 0x0
[P99332:T1:python3] ---- shim_getdents64(4, 0x8b9cf4d0, 0x8000) = 0x480
[P99332:T1:python3] ---- shim_getdents64(4, 0x8b9cf4d0, 0x8000) = 0x0
[P99332:T1:python3] ---- shim_close(4) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages/google/__init__.cpython-36m-x86_64-linux-gnu.so", 0x8d26a4c0) = -2
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages/google/__init__.abi3.so", 0x8d26a4c0) = -2
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages/google/__init__.so", 0x8d26a4c0) = -2
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages/google/__init__.py", 0x8d26a4c0) = -2
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages/google/__init__.pyc", 0x8d26a4c0) = -2
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages/google", 0x8d26a4c0) = 0x0
[P99332:T1:python3] ---- shim_read(3, 0x8bdb3cc0, 0x2000) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x0
[P99332:T1:python3] ---- shim_close(3) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3/dist-packages", 0x8d26ba90) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3/dist-packages", O_RDONLY|0x90800, 0000) = 0x3
[P99332:T1:python3] ---- shim_fstat(3, 0x8d26b880) = 0x0
[P99332:T1:python3] ---- shim_getdents64(3, 0x8b9cf4d0, 0x8000) = 0x918
[P99332:T1:python3] ---- shim_getdents64(3, 0x8b9cf4d0, 0x8000) = 0x0
[P99332:T1:python3] ---- shim_close(3) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/dist-packages", 0x8d26ba90) = -2
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6", 0x8d26a9c0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/sitecustomize.py", 0x8d26a5e0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/sitecustomize.py", 0x8d26ae80) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/__pycache__/sitecustomize.cpython-36.pyc", O_RDONLY|0x80000, 0000) = 0x3
[P99332:T1:python3] ---- shim_fstat(3, 0x8d26afe0) = 0x0
[P99332:T1:python3] ---- shim_lseek(3, 0x0, SEEK_CUR) = 0x0
[P99332:T1:python3] ---- shim_fstat(3, 0x8d26b040) = 0x0
[P99332:T1:python3] ---- shim_read(3, 0x8be80a00, 0xd5) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0xd4
[P99332:T1:python3] ---- shim_read(3, 0x8be80ad4, 0x1) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x0
[P99332:T1:python3] ---- shim_close(3) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/sitecustomize.py", O_RDONLY|0x80000, 0000) = 0x3
[P99332:T1:python3] ---- shim_fstat(3, 0x8d26afe0) = 0x0
[P99332:T1:python3] ---- shim_lseek(3, 0x0, SEEK_CUR) = 0x0
[P99332:T1:python3] ---- shim_fstat(3, 0x8d26b040) = 0x0
[P99332:T1:python3] ---- shim_read(3, 0x8bbc9110, 0x9c) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x9b
[P99332:T1:python3] ---- shim_read(3, 0x8bbc91ab, 0x1) ...
[P99332:T1:python3] ---- return from shim_read(...) = 0x0
[P99332:T1:python3] ---- shim_close(3) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/sitecustomize.py", 0x8d26ac90) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/__pycache__", 0x8d26a780) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/__pycache__/sitecustomize.cpython-36.pyc.2344022736", O_WRONLY|O_CREAT|O_EXCL|0x80000, 0600) = 0x3
[P99332:T1:python3] ---- shim_fstat(3, 0x8d26a9a0) = 0x0
[P99332:T1:python3] ---- shim_write(3, 0x8be32cd8, 0xd4) ...
[P99332:T1:python3] ---- return from shim_write(...) = 0xd4
[P99332:T1:python3] ---- shim_close(3) = 0x0
[P99332:T1:python3] ---- shim_rename("/usr/lib/python3.6/__pycache__/sitecustomize.cpython-36.pyc.2344022736", "/usr/lib/python3.6/__pycache__/sitecustomize.cpython-36.pyc") = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6", 0x8d269a80) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/lib-dynload", 0x8d269b20) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/lib-dynload", 0x8d2692c0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/lib-dynload", 0x8d269a80) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3.6/lib-dynload", O_RDONLY|0x90800, 0000) = 0x3
[P99332:T1:python3] ---- shim_fstat(3, 0x8d2699a0) = 0x0
[P99332:T1:python3] ---- shim_getdents64(3, 0x8b9cf4d0, 0x8000) = 0xa80
[P99332:T1:python3] ---- shim_getdents64(3, 0x8b9cf4d0, 0x8000) = 0x0
[P99332:T1:python3] ---- shim_close(3) = 0x0
[P99332:T1:python3] ---- shim_stat("/root/.local/lib/python3.6/site-packages", 0x8d269b20) = 0x0
[P99332:T1:python3] ---- shim_stat("/root/.local/lib/python3.6/site-packages", 0x8d2692c0) = 0x0
[P99332:T1:python3] ---- shim_stat("/root/.local/lib/python3.6/site-packages", 0x8d269a80) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/root/.local/lib/python3.6/site-packages", O_RDONLY|0x90800, 0000) = 0x3
[P99332:T1:python3] ---- shim_fstat(3, 0x8d2699a0) = 0x0
[P99332:T1:python3] ---- shim_getdents64(3, 0x8b9cf4d0, 0x8000) = 0x3d8
[P99332:T1:python3] ---- shim_getdents64(3, 0x8b9cf4d0, 0x8000) = 0x0
[P99332:T1:python3] ---- shim_close(3) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages", 0x8d269a80) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3/dist-packages", 0x8d269b20) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3/dist-packages", 0x8d2692c0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3/dist-packages", 0x8d269a80) = 0x0
[P99332:T1:python3] ---- shim_openat(AT_FDCWD, "/usr/lib/python3/dist-packages", O_RDONLY|0x90800, 0000) = 0x3
[P99332:T1:python3] ---- shim_fstat(3, 0x8d2699a0) = 0x0
[P99332:T1:python3] ---- shim_getdents64(3, 0x8b9cf4d0, 0x8000) = 0x918
[P99332:T1:python3] ---- shim_getdents64(3, 0x8b9cf4d0, 0x8000) = 0x0
[P99332:T1:python3] ---- shim_close(3) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6", 0x8d26a9c0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3.6/lib-dynload", 0x8d26a9c0) = 0x0
[P99332:T1:python3] ---- shim_stat("/root/.local/lib/python3.6/site-packages", 0x8d26a9c0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/local/lib/python3.6/dist-packages", 0x8d26a9c0) = 0x0
[P99332:T1:python3] ---- shim_stat("/usr/lib/python3/dist-packages", 0x8d26a9c0) = 0x0
[P99332:T1:python3] ---- shim_ioctl(0, TCGETS, 0x8d26d380) ...
[P99332:T1:python3] ---- return from shim_ioctl(...) = -38
[P99332:T1:python3] ---- shim_ioctl(0, TCGETS, 0x8d26d330) ...
[P99332:T1:python3] ---- return from shim_ioctl(...) = -38
[P99332:T1:python3] ---- shim_fstat(0, 0x8d26cbf0) = 0x0
[P99332:T1:python3] ---- shim_ioctl(0, TCGETS, 0x8d26cb60) ...
[P99332:T1:python3] ---- return from shim_ioctl(...) = -38
[P99332:T1:python3] ---- shim_read(0, 0x8bd1d960, 0x2000) ...
212121
[P99332:T1:python3] ---- return from shim_read(...) = 0x7
[P99332:T1:python3] ---- shim_read(0, 0x8bd1d960, 0x2000) ...
```

**Comment by mkow** [comment]
> [P99332:T1:python3] ---- shim_read(0, 0x8bd1d960, 0x2000) ...

Looks like blocked on reading from stdin?

**Comment by dimakuv** [comment]
Well, the last line in this log is a read from STDIN (standard input, or file descriptor `0`):
```
[P99332:T1:python3] ---- shim_read(0, 0x8bd1d960, 0x2000) ...
```

It looks like your TensorFlow application is waiting for some input from the user. In other words, it's not a hang, but a "user, print something and press ENTER". Isn't it normal behavior for your workload?

**Comment by duanbing** [comment]
yeah, you are right, I should give the script as an argument for entrypoint python3.  so I am confused with `sgx.trusted_files.script = "file:label_image/label_image.py"`, why this script does not be filled as the argument of the argument.  

And I run a very simple training task as below:

```
import tensorflow as tf

x = tf.Variable(3, tf.int16)
y = tf.Variable(5, tf.int16)
z = tf.add(x,y)
init = tf.global_variables_initializer()
with tf.Session() as sess:
    sess.run(init)
    print(sess.run(z))

```

It works but costs about 4 mins on my 256M-EPC server, about 100x slowdown.  That's ok, I assume I will run the training by SGX2.

**Comment by dimakuv** [comment]
> so I am confused with sgx.trusted_files.script = "file:label_image/label_image.py", why this script does not be filled as the argument of the entrypoint.

That's not what `sgx.trusted_files` is about. It does **not** fill the arguments for you. It simply tells Graphene that this particular file (`label_image/label_image.py`) should be treated as "trusted" -- this file should be hashed and its hash should be verified at runtime (to prevent the case when the attacker modified this file). See https://graphene.readthedocs.io/en/latest/manifest-syntax.html#trusted-files.

**Comment by duanbing** [comment]
Yeah, I got it. I just don't know what the `script`, or identify section means. I guess it's just an identifier, which can not be used anywhere else?

**Comment by dimakuv** [comment]
That's a good question :)

Yes, `script` is just a unique identifier. It doesn't really have any meaning for Graphene, it is only for the convenience of the manifest developer (kinda like a comment). You could as well write something like:
```
sgx.trusted_files.this_is_an_absolutely_useless_string= "file:label_image/label_image.py"
```

**Comment by duanbing** [comment]
Thanks!  @dimakuv @mkow Now I found no matter what kinda script/model I run, it costs about 6m50s to run.  Example is [here](https://github.com/duanbing/graphene/tree/master/Examples/tensorflow-training). I will try to read the source code and optimize the time cost later.  Thanks again.
